master/pkg/searcher/searcher.go

```diff
@@ -95,8 +94,7 @@ func (s *Searcher) TrialExitedEarly(
 
 	switch exitedReason {
 	case workload.InvalidHP:
-		unitsCancelled := s.UnitsCompletedByTrial[requestID]
-		s.TotalUnitsCompleted -= unitsCancelled
+		delete(s.TrialProgress, requestID)
```

Comment:
Oh, this answers my previous question.

Is there any risk of a race condition with this?  If the trial fails before the searcher state is saved?  Or is that all atomic?

Reply:
All atomic. The trial snapshot that indicates invalid hp will be saved in a transaction with the message that push us into this code path once the message is handled entirely.